<a href="https://colab.research.google.com/github/MarinaVoit/python_for_ds_tasks/blob/main/Maryna_Voit_hw_10_4_apply%2C_groupby%2C_pivot_table.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Домашнє завдання до тем apply(), groupby(), pivot_table()

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [2]:
import pandas as pd

from google.colab import drive
drive.mount('/content/drive')

df = pd.read_csv('drive/MyDrive/Colab/data/supermarket_sales.csv')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


1. Дослідимо, який філіал супермаркету ('Branch') є найприбутковішим. Для цього знайдіть сумарний прибуток за кожним філіалом і виявіть, який філіал має найвищий.

In [3]:
df.groupby('Branch')['Total'].sum().sort_values(ascending=False).head(1)

,Total
Branch,
C,110568.7065


2. В якому місті знайходиться філіал з найвищим прибутком? Може в тому місці нам розмітисти ще один магазин.  
Знайдіть відповідь за допомогою функціоналу Pandas.

In [4]:
df.groupby('City')['Total'].sum().sort_values(ascending=False).head(1)

,Total
City,
Naypyitaw,110568.7065


3.1. Створіть зводну таблицю, яка покаже, скільки покупок (інвойсів) було зроблено в кожній з філій (`Branch`) за різними категоріями товарів. Запишіть таблицю в змінну `invoices_by_category` і виведіть змінну на екран.
Ця таблиця допоможе проаналізувати, в якій філії купують найбільше товарів кожної з категорій.

In [5]:
invoices_by_category = pd.pivot_table(df, index='Branch', columns='Product line', values='Invoice ID', aggfunc='count')

invoices_by_category

Product line,Electronic accessories,Fashion accessories,Food and beverages,Health and beauty,Home and lifestyle,Sports and travel
Branch,,,,,,
A,60,51,58,47,65,59
B,55,62,50,53,50,62
C,55,65,66,52,45,45


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1rueAdko6S3UxIHGtojetTxlES-EyM6Yb)

3.2. Викристовуючи змінну `invoices_by_category` дайте відповідь програмно (тобто значення треба не просто знайти очима, а вивести за допомогою коду), в якому філіалі магазину (`Branch`) найбільше інвойсів із покупкою товарів категорії "Електронні аксесуари" (`Electronic accessories`)?


In [6]:
invoices_by_category['Electronic accessories'].sort_values(ascending=False).head(1)

,Electronic accessories
Branch,
A,60


4-6. **Творче завдання на розвиток аналітичного мислення**

Крок 1. Сформулюйте ТРИ питання (гіпотези) до наявних даних, які допомогли б вам зрозуміти, які користувачі що, де та коли найбільше/найменше купують, аби дати на основі цих гіпотез рекомендації бізнесу. Звісно питання мають бути не тими, на які ми вже відповіли в завданнях модулю.

Крок 2. Знайдіть відповіді на свої питання з допомогою функціоналу pandas.

Крок 3. Напишіть, як відповідь на це питання може бути використана для прийняття бізнес рішень.   
   
 Питання можуть бути будь-якої складності, але їх має бути 3. Кожне питання оцінюється як 1 завдання. Без виконання цього завдання ДЗ не приймається. Якщо є питання щодо виконання - пишіть у чат 🙌


In [7]:
df

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
995,233-67-5758,C,Naypyitaw,Normal,Male,Health and beauty,40.35,1,2.0175,42.3675,1/29/2019,13:46,Ewallet,40.35,4.761905,2.0175,6.2
996,303-96-2227,B,Mandalay,Normal,Female,Home and lifestyle,97.38,10,48.6900,1022.4900,3/2/2019,17:16,Ewallet,973.80,4.761905,48.6900,4.4
997,727-02-1313,A,Yangon,Member,Male,Food and beverages,31.84,1,1.5920,33.4320,2/9/2019,13:22,Cash,31.84,4.761905,1.5920,7.7
998,347-56-2442,A,Yangon,Normal,Male,Home and lifestyle,65.82,1,3.2910,69.1110,2/22/2019,15:33,Cash,65.82,4.761905,3.2910,4.1


4. У якому місяці мережа отримує найменше сумарного доходу (Total) та валового прибутку (gross income) і які дні тижня просідають?

In [8]:
df['Month'] = pd.to_datetime(df['Date']).dt.month

df.groupby('Month')[['Total', 'gross income']].sum().round(2).sort_values(by='Total')

,Total,gross income
Month,,
2,97219.37,4629.49
3,109455.51,5212.17
1,116291.87,5537.71


In [13]:
df['Day_of_week'] = pd.to_datetime(df['Date']).dt.day_name()

df.groupby('Day_of_week')['Total'].agg(
    Check_count='count',
    Mean_check='mean',
    Total_revenue='sum'
).round(2).sort_values(by='Total_revenue')

,Check_count,Mean_check,Total_revenue
Day_of_week,,,
Monday,125,303.19,37899.08
Wednesday,143,305.81,43731.14
Friday,139,316.02,43926.34
Sunday,133,334.27,44457.89
Thursday,138,328.62,45349.25
Tuesday,158,325.84,51482.25
Saturday,164,342.20,56120.81


Найменш прибутковий місяць: лютий з найнижчими показниками загального доходу та  валового прибутку. Найменш прибутковий день: понеділок, який суттєво просідає, як за кількістю чеків, так і за середнім чеком, формуючи найнижчий загальний дохід.

Використання результатів для прийняття бізнес-рішень:
- стимулювати прибуток у дні з найнижчими показниками (понеділок): запустити акцію «Щасливий понеділок» (наприклад, подвійні бали в програмі лояльності або спеціальна знижка на популярні категорії товарів);
- направляти таргетовані push-сповіщення у неділю ввечері із закликом зробити закупки на початок тижня;
- підвищити середній чек в понеділок та середу (де середній чек менший за 306 дол.) впровадженням акції за типом "купи на суму від 350 дол. та отримай подарунок/знижку".
- заздалегідь планувати маркетинговий календар на лютий (сезонні розпродажі, тематичні промо-кампанії до свят), щоб вирівняти просідання виручки порівняно із січнем та березнем.
- переглянути графіки виходу персоналу: зменшити кількість касирів та працівників залу в понеділок (коли трафік мінімальний — 125 чеків) і перенаправити ресурси на суботу (максимальний трафік — 164 чеки).

5. Як тип клієнта (Customer type) впливає на розмір середнього чека у різних категоріях товарів (Product line)?

In [19]:
customer_category_analysis = pd.pivot_table(
    df,
    index='Product line',
    columns='Customer type',
    values='Total',
    aggfunc='mean'
).round(2)

customer_category_analysis['Difference (Member - Normal)'] = (
    customer_category_analysis['Member'] - customer_category_analysis['Normal']
).round(2)

customer_category_analysis

Customer type,Member,Normal,Difference (Member - Normal)
Product line,,,
Electronic accessories,314.08,324.34,-10.26
Fashion accessories,306.09,304.15,1.94
Food and beverages,333.59,309.84,23.75
Health and beauty,353.85,295.73,58.12
Home and lifestyle,337.08,336.15,0.93
Sports and travel,324.53,340.36,-15.83


Використання результатів для прийняття бізнес-рішень:

- якщо у певній категорії середній чек звичайних покупців (Normal) вищий, ніж у учасників програми лояльності (Member), це сигнал, що клієнт не бачать цінності купувати дорогі товари з цієї категорії саме в нашій мережі. Потрібно ввести спеціальні бонуси/кешбек на ці категорії для Member(згідно таблиці категорії Electronic accessories та Sports and travel мають середній чек у клієнтів Normal вищий);

- для категорій, де Member демонструють значно вищий середній чек, варто запускати ексклюзивні попередні продажі (Early Access) чи персналізовані e-mail/push-розсилки з новинками(наприклад: в категорії Health and beauty);

- на касі можна пропонувати оформлення картки лояльності з акцентом на ті категорії, де члени клубу купують значно більше. Наприклад: «Зазвичай наші постійні клієнти платять менше за товари з категорії Health and beauty завдяки персональним бонусам — бажаєте оформити картку та заощадити вже зараз?»

6. Як дохід та середній чек відрізняється між клієнтами, а саме чоловіками (Male) та жінками (Female) залежно від категорії товарів (Product line)?

In [20]:
gender_product = pd.pivot_table(
    df,
    index='Product line',
    columns='Gender',
    values='Total',
    aggfunc=['sum', 'mean']
).round(2)

gender_product


sum              mean        
Gender                    Female      Male  Female    Male
Product line                                              
Electronic accessories  27102.02  27235.51  322.64  316.69
Fashion accessories     30437.40  23868.50  317.06  291.08
Food and beverages      33170.92  22973.93  368.57  273.50
Health and beauty       18560.99  30632.75  290.02  348.10
Home and lifestyle      30036.88  23825.04  380.21  294.14
Sports and travel       28574.72  26548.11  324.71  340.36

Жінки генерують значно вищий дохід та середній чек у категоріях: Food and beverages, Fashion accessories, Home and lifestyle.

Чоловіки суттєво переважають у категорії: Health and beauty.

Використання результатів для прийняття бізнес-рішень:

- розширити асортимент у категоріях Food and beverages, Home and lifestyle та розширити лінійку преміальних товарів для жінок, оскільки вони демонструють найвищий середній чек;
- у категорії Health and beauty розширити лінійку чоловічого догляду та спортивного харчування, оскільки саме чоловіки роблять тут основну виручку;
- для чоловіків у категоріях Food and beverages та Home and lifestyle впровадити пропозиції («готовий набір для грилю», «швидкий обід»), щоб підняти їхній низький середній чек;
- у категорії Fashion accessories для чоловіків задіяти промо (наприклад, супутні аксесуари біля каси), щоб підвищити виторг;
- налаштувати таргетовану електронну рекламу на жінок для промоції товарів для дому та продукції харчування.